host:port error

1. User enters connection information in SlamData.
2. Cannot connect.
3. User restarts.
4. Error message below prints out.

There may be a problem on the front-end, too -- in fact, I suspect there is -- please report that as well if it turns out to be the case.

Probably the API allows storage of incorrect / invalid URI data, which explodes the server on restart. The API should not allow adding mounts unless they are valid.

I will work with the user to get the structure of the Connection URI.

```
java.lang.reflect.InvocationTargetException
at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
at java.lang.reflect.Method.invoke(Method.java:606)
at com.simontuffs.onejar.Boot.run(Boot.java:340)
at com.simontuffs.onejar.Boot.main(Boot.java:166)
at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
at java.lang.reflect.Method.invoke(Method.java:606)
at com.exe4j.runtime.LauncherEngine.launch(Unknown Source)
at com.install4j.runtime.launcher.MacLauncher.main(Unknown Source)
Caused by: com.mongodb.MongoException: host and port should be specified in host:port format
at com.mongodb.ServerAddress.<init>(ServerAddress.java:122)
at com.mongodb.ServerAddress.<init>(ServerAddress.java:49)
at com.mongodb.Mongo.createCluster(Mongo.java:618)
at com.mongodb.Mongo.<init>(Mongo.java:283)
at com.mongodb.MongoClient.<init>(MongoClient.java:265)
at slamdata.engine.physical.mongodb.util$$anonfun$1.apply(util.scala:13)
at slamdata.engine.physical.mongodb.util$$anonfun$1.apply(util.scala:12)
at scalaz.MemoFunctions$$anonfun$mutableMapMemo$1$$anonfun$apply$2$$anonfun$apply$3.apply(Memo.scala:67)
at scala.collection.mutable.MapLike$class.getOrElseUpdate(MapLike.scala:194)
at scala.collection.mutable.AbstractMap.getOrElseUpdate(Map.scala:80)
at scalaz.MemoFunctions$$anonfun$mutableMapMemo$1$$anonfun$apply$2.apply(Memo.scala:67)
at slamdata.engine.physical.mongodb.util$$anonfun$2$$anonfun$apply$1.apply(util.scala:16)
at slamdata.engine.physical.mongodb.util$$anonfun$2$$anonfun$apply$1.apply(util.scala:16)
at scalaz.concurrent.Task$$anonfun$delay$1.apply(Task.scala:266)
at scalaz.concurrent.Task$$anonfun$delay$1.apply(Task.scala:266)
at scalaz.concurrent.Task$$anonfun$suspend$1$$anonfun$4.apply(Task.scala:274)
at scalaz.concurrent.Task$$anonfun$suspend$1$$anonfun$4.apply(Task.scala:274)
at scalaz.concurrent.Task$.Try(Task.scala:379)
at scalaz.concurrent.Task$$anonfun$suspend$1.apply(Task.scala:274)
at scalaz.concurrent.Task$$anonfun$suspend$1.apply(Task.scala:274)
at scalaz.concurrent.Future.step(Future.scala:111)
at scalaz.concurrent.Future.listen(Future.scala:76)
at scalaz.concurrent.Future.runAsync(Future.scala:143)
at scalaz.concurrent.Future.run(Future.scala:160)
at scalaz.concurrent.Task.run(Task.scala:97)
at slamdata.engine.api.Server$.main(server.scala:120)
at slamdata.engine.api.Server.main(server.scala)
... 12 more
```

--- a/core/src/main/scala/slamdata/engine/config/config.scala
+++ b/core/src/main/scala/slamdata/engine/config/config.scala
@@ -27,9 +27,12 @@ object Credentials {
   implicit def Codec = casecodec2(Credentials.apply, Credentials.unapply)("username", "password")
 }
 
-sealed trait BackendConfig
+sealed trait BackendConfig {
+  def validate: String \/ Unit
+}
 final case class MongoDbConfig(connectionUri: String) extends BackendConfig {
-
+  def validate =
+    MongoDbConfig.ParsedUri.unapply(connectionUri).map(κ(())) \/> ("invalid connection URI: " + connectionUri)
 }
 object MongoDbConfig {
   implicit def Codec = casecodec1(MongoDbConfig.apply, MongoDbConfig.unapply)("connectionUri")
@@ -46,7 +49,7 @@ object MongoDbConfig {
       "(?::([0-9]+))?" +     // 3: (primary) port
       "((?:,[^,/]+)*)" +     // 4: additional hosts
       "(?:/" +
-        "([^?]+)?" +         // 5: database
+        "([^/?]+)?" +        // 5: database
         "(?:\\?(.+))?" +     // 6: options
       ")?$").r
 

--- a/web/src/main/scala/slamdata/engine/api/fs.scala
+++ b/web/src/main/scala/slamdata/engine/api/fs.scala
@@ -64,7 +64,15 @@ object ResponseFormat {
   }
 }
 
-class FileSystemApi(fs: FSTable[Backend]) {
+final case class RequestError(message: String)
+object RequestError {
+  implicit def EntityEncoderRequestError: EntityEncoder[RequestError] =
+    EntityEncoder[Json].contramap[RequestError] { err =>
+      Json("error" := err.toString)
+    }
+}
+
+final case class FileSystemApi(fs: FSTable[Backend], contentPath: String, config: Config, reloader: Config => Task[Unit]) {
   import Method.{MOVE, OPTIONS}
 
   val CsvColumnsFromInitialRowsCount = 1000
@@ -162,21 +170,42 @@ class FileSystemApi(fs: FSTable[Backend]) {
   object Offset extends OptionalQueryParamDecoderMatcher[Long]("offset")
   object Limit extends OptionalQueryParamDecoderMatcher[Long]("limit")
 
-  // Note: CORS middleware is coming in http4s post-0.6.5
-  val corsHeaders = List(
-    AccessControlAllowOriginAll,
-    `Access-Control-Allow-Methods`(List("GET", "PUT", "POST", "DELETE", "MOVE", "OPTIONS")),
-    `Access-Control-Max-Age`(20*24*60*60),
-    `Access-Control-Allow-Headers`(List(Destination)))  // NB: actually needed for POST only
+  object Cors extends Middleware {
+    // Note: CORS middleware is coming in http4s post-0.6.5
+    val corsHeaders = List(
+      AccessControlAllowOriginAll,
+      `Access-Control-Allow-Methods`(List("GET", "PUT", "POST", "DELETE", "MOVE", "OPTIONS")),
+      `Access-Control-Max-Age`(20*24*60*60),
+      `Access-Control-Allow-Headers`(List(Destination)))  // NB: actually needed for POST only
 
-  // TODO: Using the Option Kleisli category or scalaz.NullResult instead of PartialFunction.
-  def corsService(pf: PartialFunction[Request, Task[Response]]) =
-    HttpService(pf.orElse[Request, Task[Response]] {
-      case OPTIONS -> _ => Ok()
-    }).map(_.putHeaders(corsHeaders: _*))
+    def apply(service: HttpService): HttpService =
+      Service.lift { req =>
+        service(req).flatMap {
+          case r@Some(_) => Task.now(r)
+          case None => Ok().map(resp => Some(resp.putHeaders(corsHeaders: _*)))
+        }
+      }
+  }
 
-  def queryService = {
-    corsService {
+  /** Handle failure in Task by returning a 500. Otherwise http4s hangs for 30 seconds and then returns 200. */
+  object FailSafe extends Middleware {
+    def apply(service: HttpService): HttpService =
+      Service.lift { req =>
+        service.run(req).handleWith {
+          case err => InternalServerError(Json("error" := err.toString)).map(Some(_))
+        }
+      }
+  }
+
+  type M[A] = EitherT[Task, RequestError, A]
+  def liftT[A](t: Task[A]): M[A] = EitherT.right(t)
+  def liftE[A](v: RequestError \/ A): M[A] = EitherT(Task.now(v))
+
+  def respond(v: M[String]): Task[Response] =
+    v.fold(BadRequest(_), Ok(_)).join
+
+  def queryService =
+    HttpService {
       case req @ GET -> AsDirPath(path) :? Q(query) =>
         (for {
           b     <- backendFor(path)
@@ -210,7 +239,6 @@ class FileSystemApi(fs: FSTable[Backend]) {
           resp <- if (query != "") go(Query(query)) else POSTContentMustContainQuery
         } yield resp
     }
-  }
 
   def compileService = {
     def go(path: Path, query: Query) =
@@ -227,7 +255,7 @@ class FileSystemApi(fs: FSTable[Backend]) {
           case PhaseResult.Detail(name, value) => Ok(name + "\n" + value)
       }).fold(identity, identity)
 
-    corsService {
+    HttpService {
       case GET -> AsDirPath(path) :? Q(query) => go(path, query)
 
       case GET -> _ => QueryParameterMustContainQuery
@@ -240,7 +268,7 @@ class FileSystemApi(fs: FSTable[Backend]) {
   }
 
   def serverService(config: Config, reloader: Config => Task[Unit]) = {
-    corsService {
+    HttpService {
       case req @ PUT -> Root / "port" => for {
         body <- EntityDecoder.decodeString(req)
         r    <- body.parseInt.fold(
@@ -258,15 +286,14 @@ class FileSystemApi(fs: FSTable[Backend]) {
   }
 
   def mountService(config: Config, reloader: Config => Task[Unit]) = {
-    def addPath(path: Path, req: Request): Task[Unit] = for {
-      body <- EntityDecoder.decodeString(req)
-      conf <- Parse.decodeEither[BackendConfig](body).fold(
-        e => Task.fail(new RuntimeException(e)),
-        Task.now)
-      _    <- reloader(config.copy(mountings = config.mountings + (path -> conf)))
+    def addPath(path: Path, req: Request): M[Unit] = for {
+      body <- liftT(EntityDecoder.decodeString(req))
+      conf <- liftE(Parse.decodeEither[BackendConfig](body).leftMap(RequestError(_)))
+      _    <- liftE(conf.validate.leftMap(RequestError(_)))
+      _    <- liftT(reloader(config.copy(mountings = config.mountings + (path -> conf))))
     } yield ()
 
-    corsService {
+    HttpService {
       case GET -> AsPath(path) =>
         config.mountings.find { case (k, _) => k.equals(path) }.fold(
           NotFound("There is no mount point at " + path))(
@@ -281,10 +308,10 @@ class FileSystemApi(fs: FSTable[Backend]) {
           case (_, None) => DestinationHeaderMustExist
         }
       case req @ POST -> AsPath(path) =>
-        def addMount = for {
-          _    <- addPath(path, req)
-          resp <- Ok("added " + path)
-        } yield resp
+        def addMount =
+          respond(for {
+            _ <- addPath(path, req)
+          } yield "added " + path)
         config.mountings.find { case (k, _) => k.contains(path) }.fold(
           addMount) {
           case (k, v) =>
@@ -295,10 +322,9 @@ class FileSystemApi(fs: FSTable[Backend]) {
               addMount
         }
       case req @ PUT -> AsPath(path) =>
-        for {
-          _    <- addPath(path, req)
-          resp <- Ok("updated " + path)
-        } yield resp
+        respond(for {
+          _ <- addPath(path, req)
+        } yield "updated " + path)
       case DELETE -> AsPath(path) =>
         if (config.mountings.contains(path))
           for {
@@ -310,7 +336,7 @@ class FileSystemApi(fs: FSTable[Backend]) {
     }
   }
 
-  def metadataService = corsService {
+  def metadataService = HttpService {
     case GET -> AsPath(path) =>
       if (path == Path("/") && fs.isEmpty)
         Ok(Json.obj("children" := List[Path]()))
@@ -388,7 +414,7 @@ class FileSystemApi(fs: FSTable[Backend]) {
       json(MediaRange.`*/*`)(DataCodec.Precise)
   }
 
-  def dataService = corsService {
+  def dataService = HttpService {
     case req @ GET -> AsPath(path) :? Offset(offset) +& Limit(limit) =>
       (for {
         _ <- offset match { case Some(o) if o < 0 => -\/(BadRequest("Negative offset: " + o)); case _ =>  \/-(()) }
@@ -438,7 +464,7 @@ class FileSystemApi(fs: FSTable[Backend]) {
   def fileMediaType(file: String): Option[MediaType] =
     MediaType.forExtension(file.split('.').last)
 
-  def staticFileService(basePath: String) = corsService {
+  def staticFileService(basePath: String) = HttpService {
     case GET -> AsPath(path) =>
       // NB: http4s/http4s#265 should give us a simple way to handle this stuff.
       val filePath = basePath + path.toString
@@ -451,8 +477,18 @@ class FileSystemApi(fs: FSTable[Backend]) {
           mt => Task.delay(resp.withContentType(Some(`Content-Type`(mt))))))
   }
 
-  def redirectService(basePath: String) = corsService {
+  def redirectService(basePath: String) = HttpService {
     case GET -> AsPath(path) =>
       TemporaryRedirect(Uri(path = basePath + path.toString))
   }
+
+  def AllServices = ListMap(
+    "/compile/fs"  -> FailSafe(Cors(compileService)),
+    "/data/fs"     -> FailSafe(Cors(dataService)),
+    "/metadata/fs" -> FailSafe(Cors(metadataService)),
+    "/mount/fs"    -> FailSafe(Cors(mountService(config, reloader))),
+    "/query/fs"    -> FailSafe(Cors(queryService)),
+    "/server"      -> FailSafe(Cors(serverService(config, reloader))),
+    "/slamdata"    -> FailSafe(Cors(staticFileService(contentPath + "/slamdata"))),
+    "/"            -> FailSafe(Cors(redirectService("/slamdata"))))
 }

--- a/web/src/main/scala/slamdata/engine/api/server.scala
+++ b/web/src/main/scala/slamdata/engine/api/server.scala
@@ -4,7 +4,6 @@ import java.io.File
 
 import slamdata.engine._
 import slamdata.engine.fp._
-import slamdata.engine.fs._
 import slamdata.engine.config._
 
 import scalaz.concurrent._
@@ -32,7 +31,7 @@ object Server {
     def restart(config: Config) = for {
       _       <- serv.fold(Task.now(()))(_.shutdown.map(ignore))
       mounted <- Mounter.mount(config)
-      server  <- run(config.server.port, mounted, contentPath, config, configPath)
+      server  <- run(config.server.port, FileSystemApi(mounted, contentPath, config, reloader(contentPath, configPath)))
       _       <- Task.delay { println("Server restarted on port " + config.server.port) }
       _       <- Task.delay { serv = Some(server) }
     } yield ()
@@ -52,23 +51,11 @@ object Server {
     } yield ()
   }
 
-  def run(port: Int, fs: FSTable[Backend], contentPath: String, config: Config, configPath: Option[String]):
-      Task[org.http4s.server.Server] = {
-    val api = new FileSystemApi(fs)
-    org.http4s.server.jetty.JettyBuilder
-      .bindHttp(port, "0.0.0.0")
-      .mountService(api.compileService,               "/compile/fs")
-      .mountService(api.dataService,                  "/data/fs")
-      .mountService(api.metadataService,              "/metadata/fs")
-      .mountService(api.mountService(config, reloader(contentPath, configPath)),
-                                                      "/mount/fs")
-      .mountService(api.queryService,                 "/query/fs")
-      .mountService(api.serverService(config, reloader(contentPath, configPath)),
-                                                      "/server")
-      .mountService(api.staticFileService(contentPath + "/slamdata"),
-                                                      "/slamdata")
-      .mountService(api.redirectService("/slamdata"), "/")
-      .start
+  def run(port: Int, api: FileSystemApi): Task[org.http4s.server.Server] = {
+    val builder = org.http4s.server.jetty.JettyBuilder.bindHttp(port, "0.0.0.0")
+    api.AllServices.toList.foldLeft(builder) {
+      case (b, (path, svc)) => b.mountService(svc, path)
+    }.start
   }
 
   private def waitForInput: Task[Unit] = {
@@ -110,7 +97,7 @@ object Server {
             config  <- Config.loadOrEmpty(options.config)
             mounted <- Mounter.mount(config)
             port = options.port.getOrElse(config.server.port)
-            server  <- run(port, mounted, options.contentPath, config, options.config)
+            server  <- run(port, FileSystemApi(mounted, options.contentPath, config, reloader(options.contentPath, options.config)))
             _       <- if (options.openClient) openBrowser(port) else Task.now(())
             _       <- Task.delay { println("Embedded server listening at port " + port) }
             _       <- Task.delay { println("Press Enter to stop.") }

--- a/web/src/test/scala/slamdata/engine/api/fs.scala
+++ b/web/src/test/scala/slamdata/engine/api/fs.scala
@@ -24,6 +24,7 @@ sealed trait Action
 object Action {
   final case class Save(path: Path, rows: List[Data]) extends Action
   final case class Append(path: Path, rows: List[Data]) extends Action
+  final case class Reload(cfg: Config) extends Action
 }
 
 class ApiSpecs extends Specification with DisjunctionMatchers with PendingWithAccurateCoverage {
@@ -40,7 +41,10 @@ class ApiSpecs extends Specification with DisjunctionMatchers with PendingWithAc
   down the server.
   */
   def withServer[A](fs: Map[Path, Backend])(body: => A): A = {
-    val srv = Server.run(port, FSTable(fs), ".", Config.empty, None).run
+    val srv = Server.run(port, FileSystemApi(FSTable(fs), ".", Config.empty, cfg => Task.delay {
+      historyBuff += Action.Reload(cfg)
+      ()
+    })).run
 
     try {
       body
@@ -967,6 +971,56 @@ class ApiSpecs extends Specification with DisjunctionMatchers with PendingWithAc
           val result = Http(path > code)
 
           result() must_== 400
+        }
+      }
+    }
+  }
+
+  "/mount/fs" should {
+    val root = svc / "mount" / "fs" / ""
+
+    "GET" should {
+      "be 404 with missing backend" in {
+        withServer(Map()) {
+          val req = root / "missing" / ""
+          val result = Http(req > code)
+
+          result() must_== 404
+        }
+      }
+    }
+
+    "PUT" should {
+      "succeed with valid MongoDB config" in {
+        withServer(Map()) {
+          val req = (root / "local").PUT
+                    .setBody("""{ "mongodb": { "connectionUri": "mongodb://localhost/test" } }""")
+          val result = Http(req OK as.String)
+
+          result() must_== "updated /local"
+          history must_== List(Action.Reload(Config(SDServerConfig(None), Map(Path("/local") -> MongoDbConfig("mongodb://localhost/test")))))
+        }
+      }
+
+      "be 400 with invalid JSON" in {
+        withServer(Map()) {
+          val req = (root / "local").PUT
+                    .setBody("""{ "mongodb":""")
+          val result = Http(req > code)
+
+          result() must_== 400
+          history must_== Nil
+        }
+      }
+
+      "be 400 with invalid MongoDB URI (extra slash)" in {
+        withServer(Map()) {
+          val req = (root / "local").PUT
+                    .setBody("""{ "mongodb": { "connectionUri": "mongodb://localhost:8080//test" } }""")
+          val result = Http(req > code)
+
+          result() must_== 400
+          history must_== Nil
         }
       }
     }

